File: scratch/extract_samples.py
import re
import json

file_path = 'external_xdean/src/tools/guobiao/core/fan.ts'
with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# Pattern to find Fan definitions
# regex = r"new Fan\(\{.*?name: '(.*?)'.*?sample: \[(.*?)\]"
# The above is hard due to multi-line...

# Let's split by "new Fan({"
parts = content.split("new Fan({")
cases = []

for part in parts[1:]:
    # Extract name
    name_match = re.search(r"name:\s*'(.*?)'", part)
    if not name_match:
        continue
    name = name_match.group(1)
    
    # Extract sample
    if 'sample:' in part:
        # Find everything inside sample: [ ... ]
        # This is tricky with nesting, but usually it's simple
        sample_start = part.find('sample: [')
        if sample_start != -1:
            # find matching bracket
            count = 0
            start = sample_start + 8
            end = start
            for i in range(start, len(part)):
                if part[i] == '[': count += 1
                elif part[i] == ']':
                    if count == 0:
                        end = i
                        break
                    else:
                        count -= 1
            sample_content = part[start:end]
            
            # Extract Hand.create(...)
            hands = re.findall(r"Hand\.create\('(.*?)'(?:,\s*\{(.*?)\})?\)", sample_content)
            for hand_str, options in hands:
                cases.append({
                    "name": name,
                    "hand": hand_str,
                    "opts": options if options else ""
                })

with open('xdean_samples.json', 'w', encoding='utf-8') as out:
    json.dump(cases, out, indent=2, ensure_ascii=False)

File: scratch/extract_samples.js
const fs = require('fs');
const filePath = 'external_xdean/src/tools/guobiao/core/fan.ts';
const content = fs.readFileSync(filePath, 'utf-8');

// Debug: check first 5000 chars
// console.log(content.substring(0, 5000));

const regex = /name:\s*['"](.+?)['"]/g;
let match;
while ((match = regex.exec(content)) !== null) {
  // console.log("Found name:", match[1]);
}

// Now try with sample
const fullRegex = /name:\s*['"](.+?)['"][\s\S]+?sample:\s*\[([\s\S]+?)\]/g;
const cases = [];
while ((match = fullRegex.exec(content)) !== null) {
    const name = match[1];
    const sampleBox = match[2];
    
    // Hand.create('...', { ... })
    const handRegex = /Hand\.create\(\s*'(.+?)'(?:\s*,\s*\{([\s\S]+?)\})?\s*\)/g;
    let handMatch;
    while ((handMatch = handRegex.exec(sampleBox)) !== null) {
        cases.push({
            name,
            hand: handMatch[1],
            opts: handMatch[2] ? handMatch[2].trim() : ""
        });
    }
}

fs.writeFileSync('xdean_samples.json', JSON.stringify(cases, null, 2));
console.log(`Extracted ${cases.length} cases.`);
